Make error

When attempting to `mix deps.get` I'm getting this error:

```
==> ex_faiss
Using cached faiss build...
g++ -I/opt/homebrew/Cellar/erlang/25.2.2/lib/erlang/erts-13.1.4/include -I/Users/jamilabreu/.cache/ex_faiss/faiss-19f7696deedc93615c3ee0ff4de22284b53e0243 -fPIC -O3 -shared -std=c++14 c_src/ex_faiss.cc c_src/ex_faiss/nif_util.cc c_src/ex_faiss/index.cc \
        c_src/ex_faiss/clustering.cc -o cache/libex_faiss.so -L/Users/jamilabreu/.cache/ex_faiss/faiss-19f7696deedc93615c3ee0ff4de22284b53e0243/build/faiss -lfaiss -flat_namespace -undefined suppress

ld: library not found for -lfaiss
collect2: error: ld returned 1 exit status
make: *** [Makefile:54: cache/libex_faiss.so] Error 1
could not compile dependency :ex_faiss, "mix compile" failed. Errors may have been logged above. You can recompile this dependency with "mix deps.compile ex_faiss", update it with "mix deps.update ex_faiss" or clean it with "mix deps.clean ex_faiss"

** (Mix) Could not compile with "make" (exit status: 2).
You need to have gcc and make installed. Try running the
commands "gcc --version" and / or "make --version". If these programs
are not installed, you will be prompted to install them.
```

```
gcc --version                                            
Apple clang version 14.0.0 (clang-1400.0.29.202)
Target: arm64-apple-darwin22.1.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

```
make --version
GNU Make 4.4
Built for aarch64-apple-darwin22.1.0
Copyright (C) 1988-2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

I'm on a new Apple M2, @seanmor5 any ideas?

Added the following to my .zshrc file:

```
export CC="/opt/homebrew/bin/gcc-12"
export CXX="/opt/homebrew/bin/g++-12"
```
Now getting this error:

```
[100%] Linking CXX shared library libfaiss.dylib
Undefined symbols for architecture arm64:
  "__Z10vpaddq_f3213__Float32x4_tS_", referenced from:
      __ZN5faiss4haddERKNS_12simd8float32ES2_ in distances_simd.cpp.o
  "__Z10vqtbl1q_u812__Uint8x16_tS_", referenced from:
      __ZNK5faiss11simd32uint814lookup_2_lanesERKS0_ in pq4_fast_scan_search_1.cpp.o
      __ZNK5faiss11simd32uint814lookup_2_lanesERKS0_ in pq4_fast_scan_search_qbs.cpp.o
  "__Z11vdupq_n_u16t", referenced from:
      __ZN5faiss12simd16uint165clearEv in pq4_fast_scan_search_1.cpp.o
      __ZN5faiss12simd16uint165clearEv in pq4_fast_scan_search_qbs.cpp.o
  "__Z8vandq_u812__Uint8x16_tS_", referenced from:
      __ZNK5faiss11simd32uint8anIS0_LDnEEES0_RKT_ in pq4_fast_scan_search_1.cpp.o
      __ZNK5faiss11simd32uint8anIS0_LDnEEES0_RKT_ in pq4_fast_scan_search_qbs.cpp.o
  "__Z9vaddq_f3213__Float32x4_tS_", referenced from:
      __ZNK5faiss12simd8float32plERKS0_ in distances_simd.cpp.o
  "__Z9vaddq_u1612__Uint16x8_tS_", referenced from:
      __ZNK5faiss12simd16uint16plERKS0_ in pq4_fast_scan_search_1.cpp.o
      __ZNK5faiss12simd16uint16plERKS0_ in pq4_fast_scan_search_qbs.cpp.o
  "__Z9vcgeq_u1612__Uint16x8_tS_", referenced from:
      __ZN5faiss6detail7simdlibL8cmp_xe32IXadL_Z9vcgeq_u1612__Uint16x8_tS3_EEEEjRK12uint16x8x2_tS6_S6_ in pq4_fast_scan_search_1.cpp.o
      __ZN5faiss6detail7simdlibL8cmp_xe32IXadL_Z9vcgeq_u1612__Uint16x8_tS3_EEEEjRK12uint16x8x2_tS6_S6_ in pq4_fast_scan_search_qbs.cpp.o
  "__Z9vcleq_u1612__Uint16x8_tS_", referenced from:
      __ZN5faiss6detail7simdlibL8cmp_xe32IXadL_Z9vcleq_u1612__Uint16x8_tS3_EEEEjRK12uint16x8x2_tS6_S6_ in pq4_fast_scan_search_1.cpp.o
      __ZN5faiss6detail7simdlibL8cmp_xe32IXadL_Z9vcleq_u1612__Uint16x8_tS3_EEEEjRK12uint16x8x2_tS6_S6_ in pq4_fast_scan_search_qbs.cpp.o
  "__Z9vmaxq_u1612__Uint16x8_tS_", referenced from:
      __ZN5faiss12simd16uint168accu_maxERKS0_ in partitioning.cpp.o
  "__Z9vminq_u1612__Uint16x8_tS_", referenced from:
      __ZN5faiss12simd16uint168accu_minERKS0_ in partitioning.cpp.o
  "__Z9vmulq_f3213__Float32x4_tS_", referenced from:
      __ZNK5faiss12simd8float32mlERKS0_ in distances_simd.cpp.o
  "__Z9vmulq_u1612__Uint16x8_tS_", referenced from:
      __ZNK5faiss12simd16uint16mlERKS0_ in pq4_fast_scan_search_1.cpp.o
      __ZNK5faiss12simd16uint16mlERKS0_ in pq4_fast_scan_search_qbs.cpp.o
  "__Z9vsubq_f3213__Float32x4_tS_", referenced from:
      __ZNK5faiss12simd8float32miERKS0_ in distances_simd.cpp.o
  "__Z9vsubq_u1612__Uint16x8_tS_", referenced from:
      __ZNK5faiss12simd16uint16miERKS0_ in pq4_fast_scan_search_1.cpp.o
      __ZNK5faiss12simd16uint16miERKS0_ in pq4_fast_scan_search_qbs.cpp.o
ld: symbol(s) not found for architecture arm64
collect2: error: ld returned 1 exit status
make[4]: *** [faiss/libfaiss.dylib] Error 1
make[3]: *** [faiss/CMakeFiles/faiss.dir/all] Error 2
make[2]: *** [faiss/CMakeFiles/faiss.dir/rule] Error 2
make[1]: *** [faiss] Error 2
make: *** [faiss] Error 2
```

I've tried just about everything in this thread. Including reading the linked wiki's and I get the same errors:

```
==> ex_faiss
cp -a /Users/name/.cache/ex_faiss/faiss-19f7696deedc93615c3ee0ff4de22284b53e0243/build/faiss cache/lib
/opt/homebrew/bin/g++ -I/opt/homebrew/Cellar/erlang/25.3/lib/erlang/erts-13.2/include -I/Users/name/.cache/ex_faiss/faiss-19f7696deedc93615c3ee0ff4de22284b53e0243 -fPIC -O3 -shared -std=c++14 c_src/ex_faiss.cc c_src/ex_faiss/nif_util.cc c_src/ex_faiss/index.cc \
		c_src/ex_faiss/clustering.cc -o cache/libex_faiss.so -Lcache/lib -lfaiss -flat_namespace -undefined suppress
make: /opt/homebrew/bin/g++: No such file or directory
make: *** [cache/libex_faiss.so] Error 1
could not compile dependency :ex_faiss, "mix compile" failed. Errors may have been logged above. You can recompile this dependency with "mix deps.compile ex_faiss", update it with "mix deps.update ex_faiss" or clean it with "mix deps.clean ex_faiss"
```
Apple M1 Pro
mac 13.3 (22E252)
```
❯ gcc -v
Apple clang version 14.0.0 (clang-1400.0.29.202)
Target: arm64-apple-darwin22.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```


I *am* able to `brew instal faiss` maybe a solution for mac's is to let us link to that instead? Faiss compilation issues are unfortunately common.

--- a/Makefile
+++ b/Makefile
@@ -40,8 +40,11 @@ ifeq ($(shell uname -s), Darwin)
 	LDFLAGS += -flat_namespace -undefined suppress
 	POST_INSTALL = install_name_tool $(EX_FAISS_CACHE_SO) -change @rpath/libfaiss.dylib @loader_path/lib/libfaiss.dylib
 
-	LLVM_PREFIX=$(shell brew --prefix llvm)
-	CMAKE_FLAGS += -DCMAKE_CXX_COMPILER=$(LLVM_PREFIX)/bin/clang++
+  ifeq ($(USE_BREW_LLVM), true)
+		LLVM_PREFIX=$(shell brew --prefix llvm)
+
+		CMAKE_FLAGS += -DCMAKE_CXX_COMPILER=$(LLVM_PREFIX)/bin/clang++
+	endif
 else
 	# Use a relative RPATH, so at runtime libex_faiss.so looks for libfaiss.so
 	# in ./lib regardless of the absolute location. This way priv can be safely
@@ -49,7 +52,7 @@ else
 	# and single quotes to escape shell variable
 	LDFLAGS += -Wl,-rpath,'$$ORIGIN/lib'
 	POST_INSTALL = $(NOOP)
-endif
+endif
 
 $(EX_FAISS_SO): $(EX_FAISS_CACHE_SO)
 	@ mkdir -p $(PRIV_DIR)
@@ -76,7 +79,7 @@ $(FAISS_LIB_DIR_FLAG):
 			git remote add origin $(FAISS_GIT_REPO) && \
 			git fetch --depth 1 origin $(FAISS_GIT_REV) && \
 			git checkout FETCH_HEAD && \
-			cmake -B build . $(CMAKE_FLAGS) && \
+		  cmake -B build . $(CMAKE_FLAGS) && \
 			make -C build -j faiss
 		touch $(FAISS_LIB_DIR_FLAG)
 

--- a/Makefile
+++ b/Makefile
@@ -40,9 +40,8 @@ ifeq ($(shell uname -s), Darwin)
 	LDFLAGS += -flat_namespace -undefined suppress
 	POST_INSTALL = install_name_tool $(EX_FAISS_CACHE_SO) -change @rpath/libfaiss.dylib @loader_path/lib/libfaiss.dylib
 
-  ifeq ($(USE_LLVM_BREW), true)
+	ifeq ($(USE_LLVM_BREW), true)
 		LLVM_PREFIX=$(shell brew --prefix llvm)
-
 		CMAKE_FLAGS += -DCMAKE_CXX_COMPILER=$(LLVM_PREFIX)/bin/clang++
 	endif
 else
@@ -79,7 +78,7 @@ $(FAISS_LIB_DIR_FLAG):
 			git remote add origin $(FAISS_GIT_REPO) && \
 			git fetch --depth 1 origin $(FAISS_GIT_REV) && \
 			git checkout FETCH_HEAD && \
-		  cmake -B build . $(CMAKE_FLAGS) && \
+			cmake -B build . $(CMAKE_FLAGS) && \
 			make -C build -j faiss
 		touch $(FAISS_LIB_DIR_FLAG)
 

--- a/Makefile
+++ b/Makefile
@@ -39,6 +39,9 @@ LDFLAGS = -L$(EX_FAISS_CACHE_LIB_DIR) -lfaiss
 ifeq ($(shell uname -s), Darwin)
 	LDFLAGS += -flat_namespace -undefined suppress
 	POST_INSTALL = install_name_tool $(EX_FAISS_CACHE_SO) -change @rpath/libfaiss.dylib @loader_path/lib/libfaiss.dylib
+
+	LLVM_PREFIX=$(shell brew --prefix llvm)
+	CMAKE_FLAGS += -DCMAKE_CXX_COMPILER=$(LLVM_PREFIX)/bin/clang++
 else
 	# Use a relative RPATH, so at runtime libex_faiss.so looks for libfaiss.so
 	# in ./lib regardless of the absolute location. This way priv can be safely
@@ -46,7 +49,7 @@ else
 	# and single quotes to escape shell variable
 	LDFLAGS += -Wl,-rpath,'$$ORIGIN/lib'
 	POST_INSTALL = $(NOOP)
-endif
+endif
 
 $(EX_FAISS_SO): $(EX_FAISS_CACHE_SO)
 	@ mkdir -p $(PRIV_DIR)

--- a/Makefile
+++ b/Makefile
@@ -40,7 +40,7 @@ ifeq ($(shell uname -s), Darwin)
 	LDFLAGS += -flat_namespace -undefined suppress
 	POST_INSTALL = install_name_tool $(EX_FAISS_CACHE_SO) -change @rpath/libfaiss.dylib @loader_path/lib/libfaiss.dylib
 
-  ifeq ($(USE_BREW_LLVM), true)
+  ifeq ($(USE_LLVM_BREW), true)
 		LLVM_PREFIX=$(shell brew --prefix llvm)
 
 		CMAKE_FLAGS += -DCMAKE_CXX_COMPILER=$(LLVM_PREFIX)/bin/clang++

--- a/README.md
+++ b/README.md
@@ -19,16 +19,13 @@ end
 ExFaiss will download, build, and cache Faiss on the first compilation. You must have CMake installed in order to build Faiss.
 
 ### MacOS Compilation
-For some reason Faiss requires OpenMP to build and the built in Apple LLVM Clang compilers so not include it. So if you want to build using a mac you will need to install LLVM from homebrew.
+For some reason Faiss requires OpenMP to build and the built in Apple LLVM Clang compilers do not include it. So if you have troubles building using a mac you will need to install LLVM from homebrew.
 
 ```shell
- 
-$ brew install llvm
+$ brew install llvm cmake
 ```
 
-And tell ExFaiss to use it by setting the environment variable `USE_BREW_LLVM=true`. 
-
-If you attempted to build once already and it failed you will need cd into the `ex_faiss` dir directly via `cd deps/ex_faise` and run `make clean`. Or `mix deps.clean ex_faiss && rm -rf ~/.cache/ex_faiss`
+And tell ExFaiss to use it by setting the environment variable `USE_LLVM_BREW=true`. 
 
 ### GPU Installation
 

--- a/README.md
+++ b/README.md
@@ -18,8 +18,9 @@ end
 
 ExFaiss will download, build, and cache Faiss on the first compilation. You must have CMake installed in order to build Faiss.
 
-### MacOS Compilation
-For some reason Faiss requires OpenMP to build and the built in Apple LLVM Clang compilers do not include it. So if you have troubles building using a mac you will need to install LLVM from homebrew.
+### macOS Compilation
+
+If you have troubles building on a macOS, you can try installing LLVM from homebrew.
 
 ```shell
 $ brew install llvm cmake

--- a/README.md
+++ b/README.md
@@ -18,6 +18,13 @@ end
 
 ExFaiss will download, build, and cache Faiss on the first compilation. You must have CMake installed in order to build Faiss.
 
+### MacOS Compilation
+For some reason ExFaiss doesn't want to build using the Apple LLVM so to make this work you will first need to install LLVM via [Homebrew](https://brew.sh/):
+
+```shell
+$ brew install llvm
+```
+
 ### GPU Installation
 
 If you have an NVIDIA GPU with CUDA installed, you can enable the GPU build by setting the environment variable `USE_CUDA=true`. Note that if you have already built Faiss without GPU support, you will need to delete the cached build before continuing. You can clean the existing installation by running `make clean`.

--- a/README.md
+++ b/README.md
@@ -19,11 +19,16 @@ end
 ExFaiss will download, build, and cache Faiss on the first compilation. You must have CMake installed in order to build Faiss.
 
 ### MacOS Compilation
-For some reason ExFaiss doesn't want to build using the Apple LLVM so to make this work you will first need to install LLVM via [Homebrew](https://brew.sh/):
+For some reason Faiss requires OpenMP to build and the built in Apple LLVM Clang compilers so not include it. So if you want to build using a mac you will need to install LLVM from homebrew.
 
 ```shell
+ 
 $ brew install llvm
 ```
+
+And tell ExFaiss to use it by setting the environment variable `USE_BREW_LLVM=true`. 
+
+If you attempted to build once already and it failed you will need cd into the `ex_faiss` dir directly via `cd deps/ex_faise` and run `make clean`. Or `mix deps.clean ex_faiss && rm -rf ~/.cache/ex_faiss`
 
 ### GPU Installation
 

